api/APIClient.test.ts

```diff
@@ -0,0 +1,418 @@
+import { z } from "zod";
+import { MockAPIImplementation, mockAPIServer } from "../test-helpers/mockAPIServer";
+import { APIClientCreator } from "./APIClient";
+import { TEST_API_URL, validateAPIClientCall } from "./TiFAPI";
+import { tifAPITransport } from "./Transport";
+import { APISchema, EndpointSchemasToFunctions, HTTPMethod } from "./TransportTypes";
+
+const TEST_BASE_URL = "http://localhost:8080"
+const TEST_ENDPOINT = "/test"
+
+const mockAPI = <T extends APISchema>({endpointSchema, endpointMocks}: {
+  endpointSchema: T,
+  endpointMocks: {
+    [EndpointName in keyof EndpointSchemasToFunctions<T>]: MockAPIImplementation<EndpointSchemasToFunctions<T>[EndpointName]>
+  }
+}) => 
+  mockAPIServer(new URL(TEST_BASE_URL), endpointSchema, endpointMocks)
+
+const apiClient = <T extends APISchema>(endpointSchema: T) => 
+  APIClientCreator(    
+    endpointSchema,
+    validateAPIClientCall,
+    tifAPITransport(TEST_API_URL)
+  )
+
+describe("MockAPIServer tests", () => {  
+  it("should throw an error when performing an unexpected request", async () => {
+    const endpointSchema = {
+      checkUser: {
+        input: {
+          query: z.object({
+            name: z.string(),
+          })
+        },
+        outputs: {
+          status204: "no-content"
+        },
+        httpRequest: {
+          method: "GET" as HTTPMethod,
+          endpoint: TEST_ENDPOINT
+        }
+      }
+    } as APISchema
+
+    mockAPI(
+      {
+        endpointSchema,
+        endpointMocks: {
+          checkUser: {
+            expectedRequest: {query: {name: "John"}} as any,
+            mockResponse: { 
+              status: 200,
+              data: undefined
+            } as unknown as never
+          }
+        }
+      }
+    )
+
+    let error = undefined;
+
+    try {
+      await apiClient(endpointSchema).checkUser({query: {name: "Johnny"}} as any)
+    } catch (e) {
+      error = e.cause.matcherResult.message.replace(/\x1B\[[0-?]*[ -/]*[@-~]/g, '');
+    }
+
+     expect(error).toEqual(
+      `expect(received).toMatchObject(expected)
+
+- Expected  - 1
++ Received  + 1
+
+  Object {
+    "query": Object {
+-     "name": "John",
++     "name": "Johnny",
+    },
+  }`
+    )
```

Comment:
Nit: Could probably use `expect(/* API Call */).rejects.toThrowError(/* Error */)`

Reply:
Needed to use that weird catching method since the error occurs in msw and not in the test. I'll add a comment for that